### PR TITLE
Sound API

### PR DESCRIPTION
**Pull request by sorenhan**
API itself is feature complete. Please see PR in SDK repository

### CODE REVIEW

File: MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs [hunk 1]
```diff
@@ -940,6 +941,129 @@ private void OnSetAnimationState(SetAnimationState payload)
...
+                }
+                if (options.Looping != null)
+                {
+                    soundInstance.loop = options.Looping.Value;
+                }
+            }
+        }
+
+
+        public static Dictionary<Guid, AudioSource> _soundInstances = new Dictionary<Guid, AudioSource>();
```
Comment (tombuMS) [inline MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs]: Underscore naming is for private fields.  Not public static.  Use s_ or no prefix.

File: MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs [hunk 2]
```diff
@@ -940,6 +941,129 @@ private void OnSetAnimationState(SetAnimationState payload)
...
+            if (index >= 0)
+            {
+                _unpausedSoundInstances.RemoveAt(index);
+            }
+        }
+        [CommandHandler(typeof(SetSoundState))]
+        private void OnSetSoundState(SetSoundState payload)
+        {
+            var actor = _actorManager.FindActor(payload.ActorId);
+            if (actor != null)
```
Comment (tombuMS) [inline MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs]: Prefer return early if null to reduce nesting

File: MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs [hunk 3]
```diff
@@ -940,6 +941,129 @@ private void OnSetAnimationState(SetAnimationState payload)
...
+                        soundInstance.time = payload.StartTimeOffset;
+                        ApplySoundStateOptions(soundInstance, payload.Options);
+                        soundInstance.Play();
+                        _soundInstances.Add(payload.Id, soundInstance);
+                        _unpausedSoundInstances.Add(payload.Id);
+                    }
+                }
+                else
+                {
+                    var soundInstance = _soundInstances[payload.Id];
```
Comment (tombuMS) [inline MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs]: Is there cases where this id coming in the payload could not be in this dictionary?  If so, this should raise a KeyNotFound exception, and should be avoided by using the contains key test around this or the TryGetValue method.

File: MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs [hunk 4]
```diff
@@ -940,6 +941,129 @@ private void OnSetAnimationState(SetAnimationState payload)
...
+                {
+                    soundInstance.loop = options.Looping.Value;
+                }
+            }
+        }
+
+
+        public static Dictionary<Guid, AudioSource> _soundInstances = new Dictionary<Guid, AudioSource>();
+        public static List<Guid> _unpausedSoundInstances = new List<Guid>();
+        int soundStoppedCheckIndex= 0;
```
Comment (tombuMS) [inline MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs]: Underscore for private fields.

File: MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/NetworkCommandPayloads.cs [hunk 5]
```diff
@@ -508,6 +508,44 @@ public class SetAnimationState : NetworkCommandPayload
...
+    /// </summary>
+    public class SetSoundState : NetworkCommandPayload
+    {
+        /// <summary>
+        /// The id of the sound instance - used to manipulate the sound after instantiation
+        /// </summary>
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// The id of the actor to reset the animation on.
```
Comment (eanders-MS) [inline MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/NetworkCommandPayloads.cs]: >The id of the actor to reset the animation on. [](start = 12, length = 46)

Comment out of date.

File: MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/NetworkCommandPayloads.cs [hunk 6]
```diff
@@ -508,6 +508,44 @@ public class SetAnimationState : NetworkCommandPayload
...
+        /// </summary>
+        public Guid ActorId { get; set; }
+
+        /// <summary>
+        /// The GUID of the sound asset to start playing
+        /// </summary>
+        public Guid SoundAssetId { get; set; }
+
+        /// <summary>
+        /// How to look at the target object.
```
Comment (eanders-MS) [inline MREUnityRuntime/MREUnityRuntimeLib/Messaging/Payloads/NetworkCommandPayloads.cs]: >How to look at the target object. [](start = 12, length = 33)

Comment out of date.

File: MREUnityRuntime/MREUnityRuntimeLib/Util/Unity/SoundFetcher.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections;
+using System.Threading.Tasks;
+using UnityEngine;
+
+namespace MixedRealityExtension.Util.Unity
+{
+    public static class SoundFetcher
```
Comment (eanders-MS) [inline MREUnityRuntime/MREUnityRuntimeLib/Util/Unity/SoundFetcher.cs]: >SoundFetcher [](start = 24, length = 12)

This is a lot of duplicated code. Can it be combined with TextureFetcher?

File: MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs (hunk 3 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs]: fixed everywhere

---
In reply to: [260559537](https://github.com/Microsoft/mixed-reality-extension-unity/pull/35#discussion_r260559537) [](ancestors = 260559537)

File: MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs (hunk 1 above)
Comment (sorenhan) [reply MREUnityRuntime/MREUnityRuntimeLib/App/MixedRealityExtensionApp.cs]: made it private.

---
In reply to: [260556109](https://github.com/Microsoft/mixed-reality-extension-unity/pull/35#discussion_r260556109) [](ancestors = 260556109)

**Pull request merged**